### PR TITLE
Idempotent writer

### PR DESCRIPTION
**Pull request by ybyzek**
### Description

Addresses https://github.com/confluentinc/event-streaming-patterns/issues/192

Idempotent Writer: https://github.com/confluentinc/event-streaming-patterns/pull/204
Idempotent Reader: https://github.com/confluentinc/event-streaming-patterns/pull/185

### New Pattern Checklist
- [x] Pattern file name should follow `docs/pattern-category/pattern-name.md`
- [x] Create an on-brand diagram (https://docs.google.com/presentation/d/1Zf256Z6fBvre3uclIbmxXsDpnTIxiBX66b13pHbGIYc/edit?usp=sharing)
- [x] Implement ksqlDB or code example (not all patterns will have code, but highly desired)
- [x] Provide references
- [x] Validate hyperlinks
- [x] Spell check (e.g. using `aspell`)

### CODE REVIEW

File: docs/event-processing/idempotent-writer.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+---
+seo:
+  title: Idempotent Writer
+  description: An Idempotent Writer produces an Event to an Event Streaming Platform exactly once.
+---
+
+# Idempotent Writer
+A writer produces [Events](../event/event.md) that are written into an [Event Stream](../event-stream/event-stream.md), and under stable conditions, each Event is recorded only once.
```
Comment (miguno) [inline docs/event-processing/idempotent-writer.md:8]: Please lowercase any occurrence of "event" unless they are linked.

File: docs/event-processing/idempotent-writer.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+However, in the case of an operational failure or a brief network outage, an [Event Source](../event-source/event-source.md) may need to retry writes. This may result in multiple copies of the same Event ending up in the Event Stream, as the first write may have actually succeeded and the client simply did not receive the response. This type of duplication is one of the perils of distributed systems.
```
Comment (miguno) [inline docs/event-processing/idempotent-writer.md:9]:
```suggestion
However, in the case of an operational failure or a brief network outage, an [Event Source](../event-source/event-source.md) may need to retry writes. This may result in multiple copies of the same Event ending up in the Event Stream, as the first write may have actually succeeded even though the producer did not receive the acknowledgement from the [Event Streaming Platform](../event-stream/event-streaming-platform.md). This type of duplication is a common failure scenario in practice and one of the perils of distributed systems.
```

File: docs/event-processing/idempotent-writer.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+## Problem
+How can an [Event Streaming Platform](../event-stream/event-streaming-platform.md) ensure that an Event Source does not write the same Event more than once?
+
+## Solution
+![idempotent-writer](../img/idempotent-writer.svg)
+
+Generally speaking, this can be addressed by native support for idempotent clients.
+This means that a writer may try to produce an Event more than once, but the Event Streaming Platform detects and discards duplicate write attempts for the same Event. The resulting Events in the Event Stream are unique, ensuring that a consumer's computed results remain accurate.
```
Comment (miguno) [inline docs/event-processing/idempotent-writer.md:18]: > ... ensuring that a consumer's computed results remain accurate.

This is not true in isolation. It only prevents producer-side duplicates from happening. The wording suggests, however, that the idempotent writer solves any issues for the consumer-side, too. This pattern, in the Considerations section, plus the text in the Idempotent Reader pattern emphasizes this bit.

File: docs/event-processing/idempotent-writer.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+## Implementation
+To make an Apache Kafka® producer idempotent, configure your producer with the following setting:
+
+```
+enable.idempotence=true
+```
+
+The Kafka producer tags each batch of Events that it sends to the Kafka cluster with a sequence number. The broker uses this sequence number to enforce deduplication of Events sent from this specific producer. Each batch's sequence number is persisted to the replicated log, so that even if the leader broker fails, any new leader will also know if a given batch is a duplicate.
```
Comment (miguno) [inline docs/event-processing/idempotent-writer.md:27]:
```suggestion
The Kafka producer tags each batch of Events that it sends to the Kafka cluster with a sequence number. Brokers in the cluster use this sequence number to enforce deduplication of Events sent from this specific producer. Each batch's sequence number is persisted to the replicated log, so that even if the leader broker fails, any new leader will also know if a given batch is a duplicate.
```

Comment (miguno) [reply docs/event-processing/idempotent-writer.md:27]: > Each batch's sequence number is persisted to the replicated log, so that even if the leader broker fails, ...

The reader will not understand what "the replicated log" refers to. Same for "leader broker". Either explain with a few words, or add links to learn more.

File: docs/event-processing/idempotent-writer.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+## Considerations
+Enabling idempotency for a Kafka producer not only ensures that duplicate Events are fenced out from the log, it also ensures that they are written in order. This is because the brokers accept a batch of Events only if its sequence number is exactly one greater than that of the last committed batch; otherwise, it results in an out-of-sequence error.
```
Comment (miguno) [inline docs/event-processing/idempotent-writer.md:30]: > fenced out from the log

I suggest to rename "the log" to "the topic".  We are normally using the term "stream" (when we talk in the abstract) or "topic" (when we talk specifically about Kafka). We rarely use the term "log", and it's confusing in this pattern as we are not explaining what "the log" actually refers to.

File: docs/event-processing/idempotent-writer.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+Exactly-once semantics (EOS) allows [Event Streaming Applications](../event-processing/event-processing-application.md) to process data without loss or duplication, ensuring that computed results are always accurate. Any solution that requires strong EOS guarantees should also enable EOS at all stages of the pipeline, not just on the writer. An Idempotent Writer is therefore typically combined with an [Idempotent Reader](../event-processing/idempotent-reader.md) and transactional processing.
```
Comment (miguno) [inline docs/event-processing/idempotent-writer.md:32]:
```suggestion
Exactly-once semantics (EOS) allow [Event Streaming Applications](../event-processing/event-processing-application.md) to process data without loss or duplication. This ensures that computed results are always consistent and accurate, even for stateful computations such as joins, [aggregations](../stream-processing/event-aggregator.md), and [windowing](../stream-processing/event-grouper.md). Any solution that requires EOS guarantees must enable EOS at all stages of the pipeline, not just on the writer. An Idempotent Writer is therefore typically combined with an [Idempotent Reader](../event-processing/idempotent-reader.md) and transactional processing.
```

**Review by miguno — CHANGES_REQUESTED** [review]
Left some comments

File: docs/event-processing/idempotent-writer.md (hunk 1 above)
Comment (ybyzek) [reply docs/event-processing/idempotent-writer.md:8]: @miguno IIUC this is in contrast to how Ben has been copy editing all the patterns.

File: docs/event-processing/idempotent-writer.md (hunk 4 above)
Comment (ybyzek) [reply docs/event-processing/idempotent-writer.md]: Updated in latest commit

File: docs/event-processing/idempotent-writer.md (hunk 5 above)
Comment (ybyzek) [reply docs/event-processing/idempotent-writer.md]: Updated in latest commit

File: docs/event-processing/idempotent-writer.md (hunk 3 above)
Comment (ybyzek) [reply docs/event-processing/idempotent-writer.md]: @miguno agree.  Removing sentence

File: docs/event-processing/idempotent-writer.md (hunk 1 above)
Comment (miguno) [reply docs/event-processing/idempotent-writer.md:8]: Then this should be changed in the copy-editing. :)

File: docs/event-processing/idempotent-writer.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+## Problem
+How can an [Event Streaming Platform](../event-stream/event-streaming-platform.md) ensure that an Event Source does not write the same Event more than once?
+
+## Solution
+![idempotent-writer](../img/idempotent-writer.svg)
+
+Generally speaking, this can be addressed by native support for idempotent clients.
+This means that a writer may try to produce an Event more than once, but the Event Streaming Platform detects and discards duplicate write attempts for the same Event.
+
+## Implementation
```
Comment (miguno) [inline docs/event-processing/idempotent-writer.md:20]: Sorry, I forgot this in my review yesterday:  Please also add ksqlDB and Kafka Streams. This can be very short, but we should tell the reader how to enable this via EOS for those two as well.

Comment (miguno) [inline docs/event-processing/idempotent-writer.md:20]: Sorry, I forgot this in my review yesterday: Please also add ksqlDB and Kafka Streams. This can be very short, but we should tell the reader how to enable this via EOS for those two as well. With links to docs.

**Review by miguno — CHANGES_REQUESTED** [review]

File: docs/event-processing/idempotent-writer.md (hunk 7 above)
Comment (ybyzek) [reply docs/event-processing/idempotent-writer.md:20]: 👍 @miguno done in latest commit

File: docs/event-processing/idempotent-writer.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+```
+enable.idempotence=true
+```
+
+The Kafka producer tags each batch of Events that it sends to the Kafka cluster with a sequence number. Brokers in the cluster use this sequence number to enforce deduplication of Events sent from this specific producer. Each batch's sequence number is persisted so that even if the [leader broker](https://www.confluent.io/blog/apache-kafka-intro-how-kafka-works/#replication) fails, the new leader broker will also know if a given batch is a duplicate.
+
+To enable [exactly-once processing guarantees](https://docs.ksqldb.io/en/latest/operate-and-deploy/exactly-once-semantics/) in ksqlDB or Kafka Streams, configure the application with the following setting, which includes enabling idempotence in the embedded producer:
+
+```
+processing.guarantee=exactly_once
```
Comment (miguno) [inline docs/event-processing/idempotent-writer.md:32]: The docs link for ksqlDB above says that it is indeed this setting.

For Kafka Streams, however, the current docs at https://docs.confluent.io/platform/current/streams/developer-guide/config-streams.html#processing-guarantee ask for a different setting:

```
processing.guarantee= exactly_once_v2
```

Comment (miguno) [reply docs/event-processing/idempotent-writer.md:32]: So, there's a bug in our documentation. The correct setting for KSTreams (as of today) is:

```
processing.guarantee=exactly_once_beta
```

The `*_v2` variant will be introduced in AK 3.0 through KIP-732.

Comment (miguno) [reply docs/event-processing/idempotent-writer.md:32]: In this case, I would stick with the v1 version of EOS for now, because it doesn't have the (current) "beta" suffix like the v2 version of EOS.

TL;DR:
```
processing.guarantee=exactly_once
```
(this gives you EOS v1)

Comment (ybyzek) [reply docs/event-processing/idempotent-writer.md:32]: @miguno to confirm, no change required to this PR?

Comment (miguno) [reply docs/event-processing/idempotent-writer.md:32]: So, no change is needed for now. We can add the `_v2` setting once it's available officially.

**Review by miguno — APPROVED** [review]

File: docs/event-processing/idempotent-writer.md (hunk 8 above)
Comment (miguno) [reply docs/event-processing/idempotent-writer.md:32]: Yes, no change is required. I already Approved subsequently.
